screenpy/actions/either.py

```diff
@@ -0,0 +1,79 @@
+"""
+try/accept logic using screenplay pattern
+"""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from screenpy.pacing import the_narrator
+from screenpy.speech_tools import get_additive_description
+
+if TYPE_CHECKING:
+    from screenpy import Actor
+    from screenpy.protocols import Performable
+
+
+class Either:
+    """Either perform followup action if the first one fails.
+
+    Allows actors to perform try/except blocks while still using screenplay pattern.
+    Examples::
+
+        the_actor.will(Either(DoAction()).or_(DoDifferentAction())
+
+        the_actor.will(
+            Either(CheckIfOnDomain(URL())).or_(Open.their_browser_on(URL())
+        )
+    """
+
+    second: tuple[Performable, ...]
+
+    def perform_as(self, the_actor: Actor) -> None:
+        """perform a try/accept using the two provided actions"""
+        # logs the first attempt even if it fails
+        # try:
+        #     with the_narrator.mic_cable_kinked():
+        #         the_actor.will(*self.first)
+        # except AssertionError:
+        #     try:
+        #         the_actor.will(*self.second)
+        #     except Exception as exc:
+        #         raise exc from None
+
+        # Since we are fully expecting assertion error to be raised we kink the
+        # cable to avoid the explanation; which only happens when the cable is not
+        # kinked
+
+        # logs the first attempt only if it succeeds.
+        with the_narrator.mic_cable_kinked():
+            try:
+                the_actor.will(*self.first)
+                return
+            except AssertionError:
```

Comment:
This is where I start leaning towards having a default but allow them to dictate which exceptions are caught because catching everything could be very troublesome.  For instance in screenpy_selenium I might want to catch WebDriverException because that's the actions I'm taking but I wouldn't want to continue to `self.second` just because I had a bad line of code that raised an `AttributeError`.

Reply:
What if the default is "everything" but we allow them to specify a shorter list?

```python
the_actor(
   Either(...).catching(DeliveryError, AssertionError).or_(...)
)
```?

Reply:
But that scenario is why i think we do need to figure out how to helpfully log the exception that was caught.

Reply:
Even the linters complain heavily when using `except Exception`  I think we should steer far away from catching everything.

Reply:
I would absolutely love to catch when the element isn't on the page.

```python
the_actor.will(Either(Click.on_the(LOGIN_BUTTON)).or_(...)
```
But not when I incorrectly called the action:
```python
the_actor.will(Either(Click.on_(LOGIN_BUTTON)).or_(...)
```

Reply:
> Even the linters complain heavily when using except Exception
Yes, that's to stop you from doing something you didn't mean to do. We know what we're trying to do here, so we can bypass the linter warnings the same way we do in [Eventually](https://github.com/ScreenPyHQ/screenpy/blob/trunk/screenpy/actions/eventually.py#L139).

> But not when I incorrectly called the action (it's not obvious but `Click` doesn't have a method `on_`)

I'm not sure if you have other examples brewing, but this one in particular will not be stopped by `Either`, as `Click.on_` will be caught before we even enter `Either`. That's just how Python works. For a case like an incorrect locator being used, that will probably be just as annoying to figure out as if we weren't using `Either`.

I'd still like to catch everything by default because that will give `Either` the greatest flexibility, and is probably the most likely thing most people will want to use it for, if you ask me. The test reader probably doesn't need to know the exact exceptions that need to be caught, just like with `Eventually`—they want to try a thing, if that fails, do a backup. 

They're probably going to be using `Either` because they've tried to make it work without, but couldn't. So they know what is going wrong, hopefully. 🤞

Reply:
> I'm not sure if you have other examples brewing

All it would take is to embed a bad call inside of an action.

```python
class CustomAction:
    def perform_as(self, actor):
        actor.will(Click.on_(LOGIN_BUTTON))
        ...

# then call that action
the_actor.will(Either(CustomAction()).or_(...)

Reply:
> I'd still like to catch everything by default because that will give Either the greatest flexibility

I can't in good conscience catch everything as a default. It risks causing more harm for testers than good.  We wouldn't prevent them from adding `Exception` to the possible excepts, but picking a minimal set of exceptions to catch should the default.  Testers can always add exceptions to ignore.  It's far more difficult to remove exceptions if testers don't know what they are.

Reply:
How about a slight compromise, then—what if we catch `AssertionError` and `ScreenPyError` by default? 

I'm just hesitant to pretty much explicitly turn this `Either` Action into only an `If` Action by default, because we only catch `AssertionError`. If i wanted to catch `TargetingError` (because the element isn't there) or `DeliveryError`, which i feel like will be common use-cases, then every time i use `Either` i'll have to add the extra line including those exceptions.

Reply:
> which i feel like will be common use-cases

I thought about that too and I agree.  The problem I keep running into is each of the screenpy packages would have a different set of defaults.  

Is this maybe a case where we sub-class `Conditionally` in each sub-package in order to setup a unique set of defaults for that package?  (like we do with `Pause` in screenpy_selenium?

Reply:
> Is this maybe a case where we sub-class Conditionally in each sub-package in order to setup a unique set of defaults for that package? (like we do with Pause in screenpy_selenium?

What if we leverage settings for this? Maybe we could have `Either` use a default provided by a setting? I wonder how hard that would be to set up...

Thanks for finding all those examples. I agree after seeing these that catching `ScreenPyError` is probably not a good idea. I was only thinking of `TargetingError` and `DeliveryError` while i was writing that suggestion—you're right, it's more complicated than that.

Reply:
Maybe for now we could just leave it as catching `AssertionError` and see how far that gets us, then we can extend it as a QOL upgrade later?

Reply:
I added the option to override the caught exceptions with the default being `AssertionError`